### PR TITLE
Preparing the v3.14 release.

### PR DESCRIPTION
**Pull request by avagin**
* Add the **read** pre-dump mode which uses process_vm_readv. (https://criu.org/Optimizing_pre_dump_algorithm)
* Use clone3() with set_tid to restore processes 
* Add --cgroup-yard option
* Use the new mount API
* Fix C/R ia32 processes on AMD #398
* Fix cross-compilation
* Many fixes here and there

### CODE REVIEW

**Comment by xemul** [comment]
> _No description provided._

@avagin , would you write a short summary, please :)

**Comment by avagin** [comment]
> @avagin , would you write a short summary, please :)

Done

File: criu/log.c [hunk 1]
```diff
@@ -199,8 +199,8 @@ void flush_early_log_buffer(int fd)
 		}
 		pos += hdr->len;
 	}
-	if (early_log_buf_off)
-		pr_warn("The early log isn't empty\n");
+	if (early_log_buf_off == EARLY_LOG_BUF_LEN)
```
Comment (xemul) [inline criu/log.c]: @0x7f454c46 , I'm a bit late, but please, drop a couple of lines about early log somewhere on the wiki. E.g. [here](https://criu.org/Logging) or a new "under-the-hood" article. Thanks :)

**Comment by xemul** [comment]
I only have one pre-requirement -- to document the --cgroup-yard on the wiki.
Well, two -- also write a summary to this PR :)

**Comment by Snorch** [comment]
At least all my patches apply cleanly according to git range-diff.

File: criu/log.c (hunk 1 above)
Comment (0x7f454c46) [reply criu/log.c]: done

File: Documentation/criu.txt [hunk 2]
```diff
@@ -266,10 +266,33 @@ For example, the command line for the above example should look like this:
     discovered automatically (usually via */proc*). This option is
     useful when one needs *criu* to skip some controllers.
 
-*--cgroup-props-ignore-default*::
-    When combined with *--cgroup-props*, makes *criu* substitute
-    a predefined controller property with the new one shipped. If the option
-    is not used, the predefined properties are merged with the provided ones.
+*--cgroup-yard* 'path'::
```
[the remark this answers is not in the document]
Comment (avagin) [reply Documentation/criu.txt]: Done

File: lib/c/criu.c [hunk 3]
```diff
@@ -987,6 +987,19 @@ int criu_local_add_cg_dump_controller(criu_opts *opts, const char *name)
 	return 0;
 }
 
+int criu_local_add_cg_yard(criu_opts *opts, const char *path)
+{
+	char *new;
+
+	new = strdup(path);
```
[the remark this answers is not in the document]
Comment (avagin) [reply lib/c/criu.c]: We don't use xstrdup in the libary code.

**Comment by adrianreber** [comment]
Would be nice to get further GCC 10 fixes from #929 (which needs a rebase). I would also like to see my cgroupv2 freezer changes in this release (if they are good enough) to continue the runc work.

File: criu/page-xfer.c [hunk 4]
```diff
@@ -480,6 +481,402 @@ static inline u32 ppb_xfer_flags(struct page_xfer *xfer, struct page_pipe_buf *p
...
+		if (bytes_read == -1) {
+			/* Handling Case 1*/
+			if (riov[start].iov_len == PAGE_SIZE) {
+				cnt_sub(CNT_PAGES_WRITTEN, 1);
+				start += 1;
+				continue;
+			} else if (errno == ESRCH) {
+				pr_debug("Target process PID:%d not found\n", pid);
+				return ESRCH;
+			}
```
Comment (mihalicyn) [inline criu/page-xfer.c]: Shouldn't we handle the rest of the error cases?
`... else { BUG_ON(1); }` as minimum

File: criu/page-xfer.c [hunk 5]
```diff
@@ -480,6 +481,402 @@ static inline u32 ppb_xfer_flags(struct page_xfer *xfer, struct page_pipe_buf *p
...
+				  unsigned long *aux_len)
+{
+	struct iovec *riov = ppb->iov;
+	ssize_t bytes_read;
+	unsigned long total_read = 0;
+	unsigned long start = 0;
+	unsigned long partial_read_bytes = 0;
+
+	while (start < ppb->nr_segs) {
+
```
Comment (mihalicyn) [inline criu/page-xfer.c]: extra line

File: criu/page-xfer.c [hunk 6]
```diff
@@ -480,6 +481,402 @@ static inline u32 ppb_xfer_flags(struct page_xfer *xfer, struct page_pipe_buf *p
...
+		aux_iov[*aux_len].iov_base = riov[*index].iov_base;
+		aux_iov[*aux_len].iov_len = riov[*index].iov_len;
+
+		(*aux_len) += 1;
+		(*index) += 1;
+	}
+
+	/* handling partially processed faulty iov*/
+	if (processed_bytes - bytes_read) {
+
```
Comment (mihalicyn) [inline criu/page-xfer.c]: extra line

File: criu/page-xfer.c [hunk 7]
```diff
@@ -480,6 +481,402 @@ static inline u32 ppb_xfer_flags(struct page_xfer *xfer, struct page_pipe_buf *p
...
+static unsigned long analyze_iov(ssize_t bytes_read, struct iovec* riov,
+				 unsigned long *index, struct iovec *aux_iov,
+				 unsigned long *aux_len)
+{
+	ssize_t processed_bytes = 0;
+	unsigned long partial_read_bytes = 0;
+
+	/* correlating iovs with read bytes */
+	while (processed_bytes < bytes_read) {
+
```
Comment (mihalicyn) [inline criu/page-xfer.c]: extra line

**Comment by mihalicyn** [comment]
IMHO If would be nice, If we also add support of symlinks to this release
https://github.com/checkpoint-restore/criu/pull/896

File: criu/page-xfer.c (hunk 4 above)
Comment (avagin) [reply criu/page-xfer.c]: No, no, no. This doesn't work this way. If you see something like this, you need to create a PR and fix this.

Comment (mihalicyn) [reply criu/page-xfer.c]: @avagin sure, I will add that to my TODO list. ;)

**Comment by avagin** [comment]
Hi all,

I have updated this PR. Could you all review it again? All additional changes will be in another PR.

File: test/jenkins/criu-fault.sh [hunk 8]
```diff
@@ -23,3 +27,4 @@ prep
 ./test/zdtm.py run -t zdtm/static/maps04 --fault 131 --keep-going --report report --pre 2:1 || fail
 ./test/zdtm.py run -t zdtm/transition/maps008 --fault 131 --keep-going --report report --pre 2:1 || fail
 ./test/zdtm.py run -t zdtm/static/maps01 --fault 132 -f h || fail
+
```
Comment (Snorch) [inline test/jenkins/criu-fault.sh]: Strange empty line in the end of file.

Comment (Snorch) [reply test/jenkins/criu-fault.sh]: https://github.com/checkpoint-restore/criu/pull/960

**Comment by avagin** [comment]
@xemul, ping

**Comment by mihalicyn** [comment]
All just fine except https://github.com/checkpoint-restore/criu/pull/961. I will prepare fixes soon (1-2 days).

**Pull request merged**